### PR TITLE
Safe Fiber races: `~combine` and `Fiber.n_any`

### PR DESCRIPTION
**Pull request by SGrondin**
Addresses #558 

I originally tried a much more complex approach before realizing the obvious one is probably good enough.

I tried to validate that this solution is sufficient by reading how fibers work internally, but I wouldn't be surprised if I missed some important detail.

The user-facing API turned out quite nice, in my opinion. Thanks for the suggestion in #558 @talex5 

This PR is a prerequisite for #584

### CODE REVIEW

**Comment by SGrondin** [comment]
@talex5 Even though this is not needed for the Executor_pool anymore, I still think it should be considered for merging. I have Lwt programs that cannot be converted to Eio without it.

**Review by talex5 — COMMENTED** [review]
Looks good. I've pushed a commit that combines the results as it goes rather than reversing the list at the end. It makes the code a bit shorter.

File: lib_eio/core/eio__core.mli [hunk 1]
```diff
@@ -209,15 +209,18 @@ module Fiber : sig
 
       If both fibers fail, {!Exn.combine} is used to combine the exceptions.
 
-      Warning: it is always possible that {i both} operations will succeed (and one result will be thrown away).
-      This is because there is a period of time after the first operation succeeds,
-      but before its fiber finishes, during which the other operation may also succeed. *)
```
[the remark this answers is not in the document]
Comment (SGrondin) [reply lib_eio/core/eio__core.mli:221]: Great catch, fixed.

**Review by talex5 — APPROVED** [review]
Squashed with some minor doc and test changes - thanks!

**Pull request merged**